File: TestCase_nArmedBandit/main_nArmedBandit.m
clear all;
close all;
clc;

global env;
global agent;

% Initialise variables for the problem
env.NBandits = 8;
randSeed = 20;
NIters = 10000;

% Initialise the environment parameters
rng(randSeed);
env.prob = rand(env.NBandits,1);
env.reward = [1 0];

agent.actionvalue = ones(env.NBandits,1);
agent.stepsize = 0.01;

a = zeros(NIters,1);
r = zeros(NIters,1);

% Run through iterations
a(1) = agent_nArmedBandit();
for i = 1:NIters-1
    [~,r(i)] = env_nArmedBandit(a(i));
    a(i+1) = agent_nArmedBandit(0,r(i),a(i));
end
[~,r(NIters)] = env_nArmedBandit(a(NIters));

% Plot results
figure;
subplot(2,2,1);
plot(1:length(a),a,'b*');
grid on;
xlabel('Iteration Number');
ylabel('Action Selected');

subplot(2,2,2);
plot(1:length(r),r,'r*');
grid on;
xlabel('Iteration Number');
ylabel('Reward Received');

subplot(2,2,3);
bar([env.prob agent.actionvalue])
grid on;
xlabel('Action');
ylabel('Action-Value');
legend('True','Inferred');

subplot(2,2,4);
avlength = 100;
for i = avlength:avlength:length(r)
    avtime(round(i/avlength)) = i;
    avval(round(i/avlength)) = mean(r(i-(avlength-1):i));
end
hold on;
grid on;
plot([0 NIters],[max(env.prob)*env.reward(1) max(env.prob)*env.reward(1)],'b--');
plot([0 NIters],[env.reward(1) env.reward(1)],'b-');
plot(avtime,avval);
xlabel('Iteration Number');
ylabel(['Averaged Reward (' num2str(avlength) ' tests)']);
legend('Average Reward','Average Maximum Reward','Absolute Maximum Reward');

File: TestCase_nArmedBandit/main_nArmedBandit.py
import numpy as np
import matplotlib.pyplot as plt
from TestCase_nArmedBandit import agent_nArmedBandit
from TestCase_nArmedBandit import env_nArmedBandit

plt.close('all')

NBandits = 8
stepsize = 0.01
randseed = 8
epsilon = 0.05

Iterations = 1000

agent = agent_nArmedBandit.Agent(NBandits, stepsize, epsilon)
env = env_nArmedBandit.Env(NBandits, randseed)

a = np.zeros(Iterations, 'int')
r = np.zeros(Iterations, 'int')

# Run through iterations
a[0] = agent.learn_and_act()
for i in range(0, (Iterations-2)):
    r[i] = env.payout(a[i])
    a[i+1] = agent.learn_and_act(r[i], a[i])
r[Iterations-1] = env.payout(a[Iterations-1])

# Visualise the results
print(env.prob)
print(agent.actionvalue)

plt.figure(0)
plt.bar(range(1, len(agent.actionvalue)+1), agent.actionvalue)
plt.title("Inferred Value Per Action", fontsize=24)
plt.xlabel("Action", fontsize=18)
plt.ylabel("Inferred Value", fontsize=18)

plt.figure(1)
plt.bar(range(1, len(env.prob)+1), env.prob)
plt.title("True Value Per Action", fontsize=24)
plt.xlabel("Action", fontsize=18)
plt.ylabel("True Value", fontsize=18)
plt.show()
